src/libstore-tests/s3-http-integration.cc

```diff
@@ -0,0 +1,283 @@
+        EXPECT_FALSE(config.cacheUri.query.empty());
+        EXPECT_EQ(config.cacheUri.query["region"], "eu-west-1");
+        EXPECT_EQ(config.cacheUri.query["endpoint"], "custom.s3.com");
+    } catch (const std::exception & e) {
+        FAIL() << "Should be able to create S3 store config: " << e.what();
+    }
+}
+
+TEST_F(S3HttpIntegrationTest, httpStoreAcceptsS3Urls)
+{
+    // Test that HttpBinaryCacheStoreConfig handles S3 URLs
+    auto schemes = S3BinaryCacheStoreConfig::uriSchemes();
+    EXPECT_TRUE(schemes.count("s3") > 0) << "S3 scheme should be supported";
+
+    // Verify HttpBinaryCacheStoreConfig doesn't directly list S3
+    auto httpSchemes = HttpBinaryCacheStoreConfig::uriSchemes();
+    EXPECT_FALSE(httpSchemes.count("s3") > 0) << "HTTP store shouldn't directly list S3 scheme";
+}
+
+TEST_F(S3HttpIntegrationTest, s3UploadRequest)
+{
+    // Test that S3 upload requests are properly configured
+    FileTransferRequest uploadReq(parseURL("s3://test-bucket/test-file"));
+    uploadReq.data = std::string("test data");
+
+    // Basic validation
+    EXPECT_EQ(uploadReq.uri.scheme, "s3");
+    EXPECT_TRUE(uploadReq.data.has_value());
+    EXPECT_EQ(*uploadReq.data, "test data");
+}
+
+TEST_F(S3HttpIntegrationTest, regionPriority)
+{
+    // Test that explicit region parameter takes precedence
+    std::string url = "s3://bucket/key?region=eu-west-1";
+    FileTransferRequest request(parseURL(url));
+
+    auto parsed = parseURL(url);
+    EXPECT_EQ(parsed.query["region"], "eu-west-1");
+}
+
+TEST_F(S3HttpIntegrationTest, customHeaders)
+{
+    // Test that custom headers are preserved through request creation
+    FileTransferRequest request(parseURL("s3://bucket/key"));
+    request.headers.emplace_back("Custom-Header", "custom-value");
+    request.headers.emplace_back("X-Test", "test-value");
+
+    // Check that custom header was added
+    bool foundCustomHeader = false;
+    bool foundTestHeader = false;
+    for (const auto & [key, value] : request.headers) {
+        if (key == "Custom-Header" && value == "custom-value") {
+            foundCustomHeader = true;
+        }
+        if (key == "X-Test" && value == "test-value") {
+            foundTestHeader = true;
+        }
+    }
+
+    EXPECT_TRUE(foundCustomHeader) << "Custom header should be preserved";
+    EXPECT_TRUE(foundTestHeader) << "Test header should be preserved";
+}
+
+TEST_F(S3HttpIntegrationTest, s3ProfileHandling)
+{
+    // Test that profile parameter is handled correctly
+    std::string url = "s3://bucket/key?profile=production";
+    FileTransferRequest request(parseURL(url));
+
+    auto parsed = parseURL(url);
+    EXPECT_EQ(parsed.query["profile"], "production");
+}
+
+TEST_F(S3HttpIntegrationTest, endpointOverride)
+{
+    // Test custom endpoint handling (e.g., for MinIO)
+    std::string url = "s3://bucket/key?endpoint=minio.local:9000&scheme=http";
+    FileTransferRequest request(parseURL(url));
+
+    auto parsed = parseURL(url);
```

Comment:
```suggestion
    auto url = parseURL("s3://bucket/key?endpoint=minio.local:9000&scheme=http");
    FileTransferRequest request(url);

```
Can skip naming the unparsed URL now, ditto in other tests